### ISSUE TITLE
Enable debugging verification diagnostics

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/dafny#1910]
Implement the LSP textDocument/codeAction request to add a "Debug" code action to diagnostics that are caused by assertions that could not be proven, and as such have a counterexample.

The "Debug" code action invokes a new `Debug` command that starts a debug server in the LSP server process, which listens to a port and implements the debug adapter protocol. More information on the debug server features can be found here: https://github.com/boogie-org/boogie/issues/539

Depends on https://github.com/boogie-org/boogie/issues/539 being implemented

**Comment by MikaelMayer** [comment]
I'm already working on verification diagnostics where I store assertions failing and their associated counter-example. Would this help to speed up this project?

**Comment by robin-aws** [comment]
I would definitely implement the debug server in Boogie rather than Dafny. The only downside will be having to track extra information specific to the Boogie client, but work we do there will also benefit other Boogie clients, such as tracking source symbols when they end up renamed for Boogie (e.g. the `iMinMinResult` example above).

> When an assertion cannot be proven within a loop, then the counterexample contains the code leading up to the loop, and an arbitrary iteration of the loop, like here:

It's only semi-arbitrary isn't it? You might have a loop invariant that only doesn't hold on the last iteration, for example. That detail should be handled by the debug server proposed in https://github.com/boogie-org/boogie/issues/539 rather than at this level though.

**Comment by keyboardDrummer** [comment]
> It's only semi-arbitrary isn't it? You might have a loop invariant that only doesn't hold on the last iteration, for example.

My understanding is that there's only one iteration from Boogie's perspective. Any cycle in the control-flow graph is transformed away by Boogie's sequence of transformations. The loop invariants specify the requires and ensures clauses for this single loop iteration.

**Comment by robin-aws** [comment]
That's true, but you want to visualize a concrete example where the invariant is not maintained. 

In your second example, if `n == 10` and `i == 0`, the invariant is true at the beginning and at the end of the loop body, so that's not a good state to visualize. `n == 1` and `i == 0` is probably the counterexample Boogie will capture instead, and something like `n == 10` and `i == 9` would be even better in the future, so the state points to the issue less ambiguously (as opposed to implying that `i == 0` is the edge case that makes it fail).

**Comment by keyboardDrummer** [comment]
> That's true, but you want to visualize a concrete example where the invariant is not maintained.
> 
> In your second example, if `n == 10` and `i == 0`, the invariant is true at the beginning and at the end of the loop body, so that's not a good state to visualize. `n == 1` and `i == 0` is probably the counterexample Boogie will capture instead, and something like `n == 10` and `i == 9` would be even better in the future, so the state points to the issue less ambiguously (as opposed to implying that `i == 0` is the edge case that makes it fail).

In my experience Boogie won't pick small numbers for the counterexample. In this example for example it picks `n == 13812` and `i == 7893`

**Comment by robin-aws** [comment]
Ah I was focussing mainly on the `i <= n` invariant that also isn't maintained. :) You're right that the line highlighted is never maintained.

I would still be careful about trying to abstract away from which loop iteration is being "debugged". For the following simplified example:

```dafny
method Bar() {
  var i := 1;
  while i < 10 
    invariant i <= 10
  {
    i := i + 2;
  }
}
```

The counterexample generated has `i == 9` on the line with the `while`.  I would lean towards keeping the accuracy of the counterexample and worry about reducing the complexity of the counterexample later on if it's feasible.

The more important point might be how to communicate the "stepping" process. I think we should be clear somehow that stepping into the loop will either skip it entirely or "break" on one of the iterations that possibly violates an assertion. I'm not yet sure how best to do that.

**Comment by keyboardDrummer** [comment]
> The counterexample generated has i == 9 on the line with the while.

I see, in this case the counterexample seems to always be the smallest `i`. But still, what does that change?

> The more important point might be how to communicate the "stepping" process. I think we should be clear somehow that stepping into the loop will either skip it entirely or "break" on one of the iterations that possibly violates an assertion. I'm not yet sure how best to do that.

I haven't seen IDE UI to indicate what the next point will be when stepping forward in a debugger. At least our debugger will allow stepping back and forth, so the programmer can step back after skipping a loop.
